File: programmers/level3/섬연결하기.py
def solution(n, costs):
    answer = 0
    parents = [i for i in range(n)]

    def find(x):
        if x != parents[x]:
            parents[x] = find(parents[x])
        return parents[x]

    def union(a, b):
        root_a = find(a)
        root_b = find(b)
        if root_a != root_b:
            if root_a > root_b:
                parents[root_a] = root_b
            else:
                parents[root_b] = root_a
            return True
        return False

    costs.sort(key=lambda x: x[2])
    for a, b, c in costs:
        if union(a, b):
            answer += c

    return answer

File: programmers/level3/섬연결하기.js
function solution(n, costs) {
  let answer = 0;
  costs.sort((a, b) => a[2] - b[2]);

  const parents = [];
  for (let i = 0; i < n; i++) {
    parents.push(i);
  }

  const find = (x) => {
    if (parents[x] === x) {
      return x;
    }
    return (parents[x] = find(parents[x]));
  };

  const union = (a, b) => {
    const pa = find(a);
    const pb = find(b);

    if (pa !== pb) {
      if (pa < pb) {
        parents[pb] = pa;
      } else {
        parents[pa] = pb;
      }
      return true;
    }
    return false;
  };

  costs.forEach(([s, e, c]) => {
    if (union(s, e)) {
      answer += c;
    }
  });

  return answer;
}
